deepmd/pt/model/model/linear_model.py

```diff
@@ -0,0 +1,230 @@
+        dp_energy = self.dp_model.forward_atomic(
+            extended_coord, extended_atype, dp_nlist
+        )["energy"].squeeze(-1)
+        # (nframes, nloc)
+        zbl_energy = self.zbl_model.forward_atomic(
+            extended_coord, extended_atype, zbl_nlist
+        )["energy"]
+
+        fit_ret = {
+            "energy": (
+                self.zbl_weight * zbl_energy + (1 - self.zbl_weight) * dp_energy
+            ).unsqueeze(-1)
+        }  # (nframes, nloc, 1)
+        return fit_ret
+
+    def serialize(self) -> dict:
+        return {
+            "dp_model": self.dp_model.serialize(),
+            "zbl_model": self.zbl_model.serialize(),
+        }
+
+    @classmethod
+    def deserialize(cls, data) -> "LinearModel":
+        dp_model = DPAtomicModel.deserialize(data["dp_model"])
+        zbl_model = PairTabModel.deserialize(data["zbl_model"])
+        return cls(dp_model, zbl_model)
+
+    def fitting_output_def(self) -> FittingOutputDef:
+        return FittingOutputDef(
+            [
+                OutputVariableDef(
+                    name="energy", shape=[1], reduciable=True, differentiable=True
+                )
+            ]
+        )
+
+    @staticmethod
+    def _compute_weight(
+        nlist: torch.Tensor,
+        rr: torch.Tensor,
+        ra: float,
+        rb: float,
+        alpha: Optional[float] = 0.1,
+    ) -> torch.Tensor:
+        """ZBL weight.
+
+        Parameters
+        ----------
+        nlist : torch.Tensor
+            the neighbour list, (nframes, nloc, nnei).
+        rr : torch.Tensor
+            pairwise distance between atom i and atom j, (nframes, nloc, nnei).
+        ra : float
+            inclusive lower boundary of the range in which the ZBL potential and the deep potential are interpolated.
+        rb : float
+            exclusive upper boundary of the range in which the ZBL potential and the deep potential are interpolated.
+        alpha : float
+            a tunable scale of the distances between atoms.
+
+        Returns
+        -------
+        torch.Tensor
+            the atomic ZBL weight for interpolation. (nframes, nloc)
+        """
+        assert (
+            rb > ra
+        ), "The upper boundary `rb` must be greater than the lower boundary `ra`."
+
+        numerator = torch.sum(
+            rr * torch.exp(-rr / alpha), dim=-1
+        )  # masked nnei will be zero, no need to handle
+        denominator = torch.sum(
+            torch.where(nlist != -1, torch.exp(-rr / alpha), torch.zeros_like(nlist)),
+            dim=-1,
+        )  # handle masked nnei.
+        sigma = numerator / denominator
+        u = (sigma - ra) / (rb - ra)
+        coef = torch.zeros_like(u)
+        left_mask = sigma < ra
+        mid_mask = (ra <= sigma) & (sigma < rb)
+        right_mask = sigma >= rb
+        coef[left_mask] = 1
+        smooth = -6 * u**5 + 15 * u**4 - 10 * u**3 + 1
+        coef[mid_mask] = smooth[mid_mask]
+        coef[right_mask] = 0
```

Comment:
please double check this part.

Reply:
It is recommended to check against the corresponding tf operator 
https://github.com/deepmodeling/deepmd-kit/blob/cd77429dcbdaf42892ac3319b6fb72b9a2879a65/deepmd/tf/model/ener.py#L246-L256